googlesource/source.go

```diff
@@ -17,41 +17,63 @@ package googlesource
 import (
 	"context"
 	"errors"
+	"fmt"
 	"sync"
 	"time"
 
 	"cloud.google.com/go/bigquery"
 	sdk "github.com/conduitio/conduit-connector-sdk"
 	googlebigquery "github.com/neha-Gupta1/conduit-connector-bigquery"
-	bqStoragepb "google.golang.org/genproto/googleapis/cloud/bigquery/storage/v1"
+	"google.golang.org/api/option"
 	"gopkg.in/tomb.v2"
 )
 
 type Source struct {
 	sdk.UnimplementedSource
-	Session         *bqStoragepb.ReadSession
-	BQReadClient    *bigquery.Client
-	SourceConfig    googlebigquery.SourceConfig
-	Tables          []string
-	Ctx             context.Context
-	SDKResponse     chan sdk.Record
-	LatestPositions latestPositions
-	Position        Position
-	ticker          *time.Ticker
-	tomb            *tomb.Tomb
-	iteratorClosed  chan bool
+	bqReadClient *bigquery.Client
+	sourceConfig googlebigquery.SourceConfig
+	// if a user specified table IDs in the config, then the tables here will be those same table IDs.
+	//  However, if the table IDs in config are not specified, then the tables here will contain the list of all tables
+	tables []string
+	//  for all the function running in goroutine we needed the ctx value. To provide the current
+	// ctx value ctx was required in struct.
+	ctx            context.Context
+	records        chan sdk.Record
+	positions      positions
+	ticker         *time.Ticker
+	tomb           *tomb.Tomb
+	iteratorClosed bool
+	// interface to provide BigQuery client. In testing this will be used to mock the client
+	clientType clientI
 }
 
-type latestPositions struct {
-	LatestPositions map[string]Position
-	lock            sync.Mutex
+// positions struct to maintain syncing status of tables
+type positions struct {
+	// position - map of tableIds and corresponding last position synced
+	positions map[string]string
+	// lock for goroutine safe access of position
+	lock *sync.Mutex
 }
 
-type Position struct {
+type Key struct {
 	TableID string
-	Offset  int
+	Offset  string
 }
 
+// clientI provides function to create BigQuery Client
+type clientI interface {
+	Client() (*bigquery.Client, error)
+}
```

Comment:
What we have here is actually a factory pattern, i.e. `clientI` is not a client, but a client factory, so we should call it in such a way.

Reply:
I am trying to mock the create client function itself and not the subsequent call. Creating the `client interface` will not do that task. Also, I am handling these conditions in integration testing.

Reply:
That's true, a client interface on its own, won't help you control the way a client is control, and I do agree that a "create client" function is needed. That's why I recommended renaming `clientI` to `clientFactory`, because it doesn't sound right that a client is able to create a client.

> Also, I am handling these conditions in integration testing.

No doubt, that's a great thing, and gives most confidence. Having a client interface though, makes it possible to test some cases you can't have in integration tests (e.g. API returns 429 Too Many Requests). It also hides some of the complexity (e.g. the source asks the BigQuery to execute the query, and it prepares the query string etc. all under the hood). If we ever need to change the client, it would be easy to do as well. If we agree to have a client interface, we can do it separately.

Reply:
Created one - https://github.com/neha-Gupta1/conduit-connector-bigquery/issues/27

Reply:
:+1: